Source/WebCore/rendering/RenderLayer.cpp

```diff
@@ -5837,6 +5842,9 @@ void RenderLayer::styleChanged(StyleDifference diff, const RenderStyle* oldStyle
             if (visibilityChanged || oldStyle->isOverflowVisible() != renderer().style().isOverflowVisible())
                 m_scrollableArea->computeHasCompositedScrollableOverflow(diff <= StyleDifference::RepaintLayer ? LayoutUpToDate::Yes : LayoutUpToDate::No);
         }
+
+        if (!oldStyle->opacity() != !renderer().style().opacity())
```

Comment:
That looks odd, I'd rewrite it without the negations.

`if (oldStyle->isOpaque() != style().isOpaque())` and introduce a new `bool isOpaque() const` helper in RenderStyle, that, unlike `hasOpacity` exactly checks for opacity == 0.